### ISSUE TITLE
Deploying OVF exported by vsphere-iso results in VM with empty disks

### ISSUE THREAD

**Issue by nilic** [hashicorp/packer#9629]
Initially a part of #8995, now opening a separate issue as requested.

The OVF exported by the `vsphere-iso` builder currently seems unusable.

After #9020 was released as a part of Packer 1.5.6 I tried the following:

- deploying the exported OVF to two vCenter servers using the vsphere post-processor

- converting the OVF to OVA manually using ovftool and uploading it to vCloud Director

All of these attempts have the same result - I get a VM which seemingly looks fine (in terms of virtual hardware, disk size etc) but has virtual disks which are zero in size:

![80977876-a2885800-8e25-11ea-93d8-afb4c5ebdb7c](https://user-images.githubusercontent.com/6827325/88169549-68587680-cc1c-11ea-84e8-091ef0ac13d4.png)

and doesn't boot.

Message I get from ovftool when converting to OVA:

```
Wrong file size specified in OVF descriptor for 'centos7-XXX-disk-0.vmdk' (specified: 12884901888, actual 1344942592).
Wrong file size specified in OVF descriptor for 'centos7-XXX-disk-1.vmdk' (specified: 12884901888, actual 74752).
```

Virtual disk 0 is 10 GB and disk 1 is 2 GB in size if that's of any help.

Packer 1.5.6
ovftool 4.3.0 (build-13981069)
vSphere 6.5

**Comment by jhawk28** [comment]
Can you post a sanitized version of your packer template?

**Comment by nilic** [comment]
Here:

```json
{
  "variables": {
    "build_id": "centos7",
    "vmware_guest_os": "rhel7_64Guest",
    "iso_url": "http://centos.petarmaric.com/7.7.1908/isos/x86_64/CentOS-7-x86_64-DVD-1908.iso",
    "iso_checksum_url": "http://centos.petarmaric.com/7.7.1908/isos/x86_64/sha256sum.txt",
    "build_time": "{{isotime \"060102-1504\"}}",
    "build_date": "{{isotime \"02.01.2006\"}}",
    "template_name": "{{user `build_id`}}-{{user `build_time`}}",
    "vcenter_username": "{{env `VCENTER_USERNAME`}}",
    "vcenter_password": "{{env `VCENTER_PASSWORD`}}",
    "vcd_username": "{{env `VCD_USERNAME`}}",
    "vcd_password": "{{env `VCD_PASSWORD`}}"
  },
  "sensitive-variables": [ 
    "vcenter_username",
    "vcenter_password",
    "vcd_username",
    "vcd_password"
  ],
  "builders": [
    {
      "type": "vsphere-iso",
      "name": "{{user `build_id`}}",
      "vcenter_server": "{{user `build_vcenter`}}",
      "username": "{{user `vcenter_username`}}",
      "password": "{{user `vcenter_password`}}",
      "insecure_connection": true,
      "host": "{{user `build_host`}}",
      "datastore": "{{user `build_datastore`}}",
      "folder": "{{user `build_folder`}}",
      "vm_name": "{{user `template_name`}}",
      "guest_os_type": "{{user `vmware_guest_os`}}",
      "vm_version": "11",
      "CPUs": 1,
      "cpu_cores": 1,
      "CPU_hot_plug": true,
      "RAM": 1024,
      "RAM_hot_plug": true,
      "network": "{{user `build_portgroup`}}",
      "network_card": "vmxnet3",
      "disk_controller_type": "pvscsi",
      "disk_size":        10240,
      "disk_thin_provisioned": true,
      "storage": [
        {
          "disk_size":        2048,
          "disk_thin_provisioned": true
        }
      ],
      "cdrom_type": "ide",
      "iso_urls": [
        "{{user `iso_url`}}"
      ],
      "iso_checksum_url": "{{user `iso_checksum_url`}}",
      "iso_checksum_type": "sha256",
      "http_directory" :  "{{template_dir}}/http",
      "boot_command": [
        "<esc><wait>",
        "linux inst.ks=http://{{ .HTTPIP }}:{{ .HTTPPort }}/ks.cfg",
        "<enter>"
      ],
      "communicator": "ssh",
      "ssh_username": "secretuser",
      "ssh_password": "secretpass",
      "ssh_timeout": "60m",
      "shutdown_command": "/sbin/shutdown -h now",
      "shutdown_timeout": "1000s",
      "export": {
        "images": false,
        "output_directory": "{{pwd}}/output-{{build_name}}"
      }
    }
  ],
  "provisioners": [
   ...
  ],
  "post-processors": [
    {
      "type": "shell-local",
      "name": "export-tmp-template-name_{{build_name}}",
      "command": "echo TMP_TEMPLATE_NAME={{user `template_name`}} > {{template_dir}}/../../tmp_template_name"
    },
    [
      {
        "type": "vsphere",
        "name":"upload-{{build_name}}_{{user `target_vcenter_1`}}",
        "host": "{{user `target_vcenter_1`}}",
        "username": "{{user `vcenter_username`}}",
        "password": "{{user `vcenter_password`}}",
        "insecure": true,
        "vm_name": "{{user `template_name`}}",
        "cluster": "{{user `target_cluster_1`}}",
        "datacenter": "{{user `target_datacenter_1`}}",
        "vm_folder": "{{user `target_folder_1`}}",
        "datastore": "{{user `target_datastore_1`}}",
        "disk_mode": "thin",
        "vm_network": "{{user `target_network_1`}}",
        "keep_input_artifact": true
      },
      {
        "type": "vsphere-template",
        "name": "convert-to-template-{{build_name}}_{{user `target_vcenter_1`}}",
        "host": "{{user `target_vcenter_1`}}",
        "username": "{{user `vcenter_username`}}",
        "password": "{{user `vcenter_password`}}",
        "insecure": true,
        "datacenter": "{{user `target_datacenter_1`}}",
        "folder": "{{user `target_folder_1`}}",
        "reregister_vm": false
      }
    ],
    [
      {
        "type": "vsphere",
        "name":"upload-{{build_name}}_{{user `target_vcenter_2`}}",
        "host": "{{user `target_vcenter_2`}}",
        "username": "{{user `vcenter_username`}}",
        "password": "{{user `vcenter_password`}}",
        "insecure": true,
        "vm_name": "{{user `template_name`}}",
        "cluster": "{{user `target_cluster_2`}}",
        "datacenter": "{{user `target_datacenter_2`}}",
        "vm_folder": "{{user `target_folder_2`}}",
        "datastore": "{{user `target_datastore_2`}}",
        "disk_mode": "thin",
        "vm_network": "{{user `target_network_2`}}",
        "keep_input_artifact": true
      },
      {
        "type": "vsphere-template",
        "name": "convert-to-template-{{build_name}}_{{user `target_vcenter_2`}}",
        "host": "{{user `target_vcenter_2`}}",
        "username": "{{user `vcenter_username`}}",
        "password": "{{user `vcenter_password`}}",
        "insecure": true,
        "datacenter": "{{user `target_datacenter_2`}}",
        "folder": "{{user `target_folder_2`}}",
        "reregister_vm": false
      }
    ],
    ...
  ]
}
```

**Comment by arogarth** [comment]
Similiar her - Packer 1.6.x and vCenter 6.5

But, the VMDK hast expected size in storage Browser bu on Boot the disk is empty.

My Workflow:
* Build VM using vsphere-iso
* Extract VM
* Upload VM to http storage using curl
* Deploy Image to vcenter using ovftool from URL to vCenter

This results in a failed VM.

Confusing - if i Import the Same URL using UI - it works :/ - The Same if I upload the OVF from local to vcenter

If I extract the Builded VM using ovftool and than upload it it vcenter - it works...

**Comment by nilic** [comment]
Any updates?

**Comment by smallsam** [comment]
Just spotted this behaviour too. Apparently it's an issue with vCenter 6.5/ovftool.
https://communities.vmware.com/thread/598619
https://github.com/hashicorp/packer/issues/6742

Workarounds for ESXi 6.5 appear to be removing the vmw:BootOrderSection section from the ovf file and what I'm just about to try is adding this to the packer config to prevent it ending up in there in the first place.

`      "vmx_data_post": {
        "bios.hddorder": ""
      },
`

**Comment by SwampDragons** [comment]
Interesting. We documented this for the vsphere-template post-processor: https://www.packer.io/docs/post-processors/vsphere-template#troubleshooting but I didn't know it was also a problem for the vsphere-iso builder. Makes sense since the underlying issue is the ofvtool. 

Sounds like since this is an upstream bug the best thing we can do is document it on the vsphere-iso builder too.

**Comment by smallsam** [comment]
So there isn't vmx_data_post for the vsphere iso builder.

@nilic for reference, I have the following workaround to remove the offending section from the ovf file before packing it into an ova:

`  "post-processors": [
    [
      {
        "type": "shell-local",
        "inline": [
          "xmlstarlet ed -P --inplace -d './/vmw:BootOrderSection' vm.ovf"
        ]
      },
      {
        "type": "shell-local",
        "inline": [
          "cd vmware",
          "openssl sha256 vm*.vmdk vm*.nvram vm.ovf > vm.mf"
        ]
      },      
     {
        "type": "shell-local",
        "inline": [
          "ovftool vm.ovf vm.ova"
        ]
      }
    ]`

**Comment by emaayan** [comment]
hi, 
i've also encountered this issue, i'm using vsphere-iso to create a vm on esxi 6.5, via vcenter 6.7, but the thing is , that if export if from vsphere ui, the ovf doesn't include the offending boot  section, neither does using ovftool 4.4, and exporting it via vsphere-iso does, so, what's the difference?

**Comment by jhawk28** [comment]
@emaayan the closest exporter would be govc.

**Comment by emaayan** [comment]
> @emaayan the closest exporter would be govc.

not sure i understand.

**Comment by emaayan** [comment]
> hi,
> i've also encountered this issue, i'm using vsphere-iso to create a vm on esxi 6.5, via vcenter 6.7, but the thing is , that if export if from vsphere ui, the ovf doesn't include the offending boot section, neither does using ovftool 4.4, and exporting it via vsphere-iso does, so, what's the difference?

do you recommend using govc over ovftool for exporting the OVA?

**Comment by emaayan** [comment]
> Interesting. We documented this for the vsphere-template post-processor: https://www.packer.io/docs/post-processors/vsphere-template#troubleshooting but I didn't know it was also a problem for the vsphere-iso builder. Makes sense since the underlying issue is the ofvtool.
> 
> Sounds like since this is an upstream bug the best thing we can do is document it on the vsphere-iso builder too.

would it be possible for an option to suppress the vmw:BootOrderSection in vsphere-iso builder? (because we can't change the vmx data directly) perhaps add an option "NONE" to the boot_order configuration option ? (cause setting boot_order to cdrom, still causes that element to show up.

**Comment by jhawk28** [comment]
@emaayan the vsphere-iso exporter uses the same logic/code as govc. If govc works, then it means that we can do something to fix it. If govc break, then there is an upstream bug that we should submit to VMware.

**Comment by jhawk28** [comment]
The ovf export is using the same vm object that it used to create the vm, but it could be "dirty". If govc exports the vm correctly, then if we re-get the vm object instead of using the one we "created" may give us a clean version that doesn't have the boot order. 

The original issue is that it is exporting 0KB disks. This may be caused by zeroing out the free space. We may be able to safely just not add the disk files if their download size is '0'.

**Comment by emaayan** [comment]
from what i saw govc export the vm that esxi's can import them, but when i
tried to import them via vcenter 6.7 to esxi 6.5 i also had issues of
timeouts..

**Comment by jhawk28** [comment]
This indicates that getting a clean vm object would generate the correct manifest if the govc export worked. You can verify by seeing if the boot order section was missing from the govc version and present in the vsphere-iso version.

> from what i saw govc export the vm that esxi's can import them, but when i tried to import them via vcenter 6.7 to esxi 6.5 i also had issues of timeouts..

**Comment by jhawk28** [comment]
I tried "cleaning" the boot order before export and it didnt remove the offending section in the OVF.
